### PR TITLE
reduce implementation code, API is unchanged

### PR DESCRIPTION
**Pull request by mxmlnkn**
This pull request reduces code by redirecting it to common functions. For the 4 constructors all calls are redirected to the private method `createNewPngwriter`. The API and behavior should be completely unchanged by this commit. It's only about code style.

One exception is 86d38122b809cbed23da75ce636f3b0adffea82f This commit actually changes something in order to solve a possible free on unallocated memory. In the first place null pointer dereferencing was also possible.

Following are commits and notes which prove that the changed codes really were identical at heart.

``` diff
1c1
diff "../src/constructor-iidconstchar.txt" "../src/constructor-iiiconstchar.txt" 
< pngwriter::pngwriter(int x, int y, double backgroundcolour, const char * filename)

---
> pngwriter::pngwriter(int x, int y, int backgroundcolour, const char * filename)
4a5
>    backgroundcolour_ = backgroundcolour;
6d6
<    backgroundcolour_ = int(backgroundcolour*65535);
25d24
<   width_ = 1;
26a26
>   width_ = 1;
79,80c79,80
<        //etc
<        tempindex = 6*hhh;

---
>        //graph_[vhhh][6*hhh + i] where i = 0 to 5
>        tempindex=6*hhh;
```

Note: The line with `width_` is only switch up with the line for `height_`, the other difference are whitespaces and a comment

``` diff
diff "../src/constructor-iidconstchar.txt" "../src/constructor-iidchar.txt" 
1c1
< pngwriter::pngwriter(int x, int y, double backgroundcolour, const char * filename)

---
> pngwriter::pngwriter(int x, int y, double backgroundcolour, char * filename)
6d5
<    backgroundcolour_ = int(backgroundcolour*65535);
8a8
>    backgroundcolour_ = int(backgroundcolour*65535);
25d24
<   height_ = 1;
26a26
>   height_ = 1;
31c31
<   std::cerr << " PNGwriter::pngwriter - WARNING **: Constructor called with background colour greater than 65535. Setting to 65535."<<std::endl;

---
>   std::cerr << " PNGwriter::pngwriter - WARNING **: Constructor called with background colour greater than 1.0. Setting to 1.0."<<std::endl;
35c35
<    if(backgroundcolour_ <0)

---
>    if(backgroundcolour_ < 0)
37c37
<   std::cerr << " PNGwriter::pngwriter - WARNING **: Constructor called with background colour lower than 0. Setting to 0."<<std::endl;

---
>   std::cerr << " PNGwriter::pngwriter - WARNING **: Constructor called with background colour lower than 0.0. Setting to 0.0."<<std::endl;
79,80c79,80
<        //etc
<        tempindex=6*hhh;

---
>        // graph_[vhhh][tempindex + i] where i = 0 to 5
>        tempindex = 6*hhh;
```

Note: the `backgroundcolor_` line is only switched up with 

```
   filegamma_ = 0.6;
   transformation_ = 0;
```

so it is completely unimportant, then there are only white space errors and the messages differe in 1.0 vs. 65535 and 0 vs. 0.0

``` diff
diff "../src/constructor-iidchar.txt" "../src/constructor-iiichar.txt"
1c1
< pngwriter::pngwriter(int x, int y, double backgroundcolour, char * filename)

---
> pngwriter::pngwriter(int x, int y, int backgroundcolour, char * filename)
4a5
>    backgroundcolour_ = backgroundcolour;
8d8
<    backgroundcolour_ = int(backgroundcolour*65535);
31c31
<   std::cerr << " PNGwriter::pngwriter - WARNING **: Constructor called with background colour greater than 1.0. Setting to 1.0."<<std::endl;

---
>   std::cerr << " PNGwriter::pngwriter - WARNING **: Constructor called with background colour greater than 65535. Setting to 65535."<<std::endl;
35c35
<    if(backgroundcolour_ < 0)

---
>    if(backgroundcolour_ <0)
37c37
<   std::cerr << " PNGwriter::pngwriter - WARNING **: Constructor called with background colour lower than 0.0. Setting to 0.0."<<std::endl;

---
>   std::cerr << " PNGwriter::pngwriter - WARNING **: Constructor called with background colour lower than 0. Setting to 0."<<std::endl;
79c79
<        // graph_[vhhh][tempindex + i] where i = 0 to 5

---
>        //graph_[vhhh][6*hhh + i] i = 0  to 5
```

Note: again no differences of value

``` diff
diff "../src/constructor-default.txt" "../src/constructor-iiichar.txt"
1c1
< pngwriter::pngwriter()

---
> pngwriter::pngwriter(int x, int y, int backgroundcolour, char * filename)
2a3,8
>    width_ = x;
>    height_ = y;
>    backgroundcolour_ = backgroundcolour;
>    compressionlevel_ = -2;
>    filegamma_ = 0.6;
>    transformation_ = 0;
4d9
<    filename_ = new char[255];
7c12
<    texttitle_  = new char[255];

---
>    texttitle_ = new char[strlen(filename)+1];
9,16c14
< 
<    strcpy(filename_, "out.png");
<    width_ = 250;
<    height_ = 250;
<    backgroundcolour_ = 65535;
<    compressionlevel_ = -2;
<    filegamma_ = 0.5;
<    transformation_ = 0;

---
>    filename_ = new char[strlen(filename)+1];
21c19,39
<    strcpy(texttitle_, "out.png");

---
>    strcpy(texttitle_, filename);
>    strcpy(filename_, filename);
> 
>    if((width_<0)||(height_<0))
>      {
>   std::cerr << " PNGwriter::pngwriter - ERROR **: Constructor called with negative height or width. Setting width and height to 1." << std::endl;
>   width_ = 1;
>   height_ = 1;
>      }
> 
>    if(backgroundcolour_ >65535)
>      {
>   std::cerr << " PNGwriter::pngwriter - WARNING **: Constructor called with background colour greater than 65535. Setting to 65535."<<std::endl;
>   backgroundcolour_ = 65535;
>      }
> 
>    if(backgroundcolour_ <0)
>      {
>   std::cerr << " PNGwriter::pngwriter - WARNING **: Constructor called with background colour lower than 0. Setting to 0."<<std::endl;
>   backgroundcolour_ = 0;
>      }
48a67,73
>    if(backgroundcolour_ == 0)
>      for(int vhhh = 0; vhhh<height_;vhhh++)
>        memset( graph_[vhhh],
>                (char) backgroundcolour_,
>                width_*6 );
>    else
>    {
54c79
<        //graph_[vhhh][6*hhh + i] where i goes from 0 to 5

---
>        //graph_[vhhh][6*hhh + i] i = 0  to 5
63a89
>    }
```

Note: the differences basically boild down to default parameters:
  x = 250, y = 250, backgroundcolour = 65535, filename = "out.png"
For some weird reason the fileGamme is 0.5 instead of 0.6. I don't think this was intended, so I'll neglect this difference. Also all the backgroundcolor checks are missing, but that is not significant either

### CODE REVIEW

**Comment by mxmlnkn** [comment]
Comments about the new commits:
```diff
diff ../src/copy-assignment.txt ../src/copy-constructor.txt 
1c1
< pngwriter & pngwriter::operator = (const pngwriter & rhs)
---
> pngwriter::pngwriter(const pngwriter &rhs)
3,8d2
<    if( this==&rhs)
<       return *this;
< 
<    // free old allocations from member variables
<    deleteMembers();
< 
14c8
<    transformation_ = rhs.transformation_;
---
>    transformation_ = rhs.transformation_;;
21a16
>    strcpy(filename_, rhs.filename_);
26d20
<    strcpy(filename_, rhs.filename_);
53d46
< 
58a52
> 	     //   graph_[vhhh][6*hhh + i ] i=0 to 5
69,70c63
<    return *this;
< }
---
> };
```

```diff
diff ../src/scale_kxky.txt ../src/scale_k.txt 
1c1
< void pngwriter::scale_kxky(double kx, double ky)
---
> void pngwriter::scale_k(double k)
3c3
<    if((kx <= 0.0)||(ky <= 0.0))
---
>    if(k <= 0.0)
5c5
< 	std::cerr << " PNGwriter::scale_kxky - ERROR **:  scale_kxky() called with negative or zero scale factor. Was: " << kx << ", " << ky << "." << std::endl;
---
> 	std::cerr << " PNGwriter::scale_k - ERROR **:  scale_k() called with negative or zero scale factor. Was: " << k << "." << std::endl;
7a8
>    // Calculate the new scaled height and width
9,10c10,11
<    scaledw = (int) ceil(kx*width_);
<    scaledh = (int) ceil(ky*height_);
---
>    scaledw = (int) ceil(k*width_);
>    scaledh = (int) ceil(k*height_);
12c13,14
<    pngwriter temp(scaledw, scaledh, 0, "temp");
---
>    // Create image storage.
>    pngwriter temp(scaledw,scaledh,0,"temp");
33a36
> 
53c56
< 	std::cerr << " PNGwriter::scale_kxky - ERROR **:  Not able to allocate memory for image." << std::endl;
---
> 	std::cerr << " PNGwriter::scale_k - ERROR **:  Not able to allocate memory for image." << std::endl;
61c64
< 	     std::cerr << " PNGwriter::scale_kxky - ERROR **:  Not able to allocate memory for image." << std::endl;
---
> 	     std::cerr << " PNGwriter::scale_k - ERROR **:  Not able to allocate memory for image." << std::endl;
67c70
< 	std::cerr << " PNGwriter::scale_kxky - ERROR **:  Not able to allocate memory for image." << std::endl;
---
> 	std::cerr << " PNGwriter::scale_k - ERROR **:  Not able to allocate memory for image." << std::endl;
89d91
<    //
```

```diff
diff ../src/scale_kxky.txt <(sed 's/finalwidth/scaledw/g; s/finalheight/scaledh/g;' ../src/scale_wh.txt)
1c1
< void pngwriter::scale_kxky(double kx, double ky)
---
> void pngwriter::scale_wh(int scaledw, int scaledh)
3c3
<    if((kx <= 0.0)||(ky <= 0.0))
---
>    if((scaledw <= 0)||(scaledh <= 0))
5c5
< 	std::cerr << " PNGwriter::scale_kxky - ERROR **:  scale_kxky() called with negative or zero scale factor. Was: " << kx << ", " << ky << "." << std::endl;
---
> 	std::cerr << " PNGwriter::scale_wh - ERROR **: Negative or zero final width or height not allowed." << std::endl;
8,11d7
<    int scaledh, scaledw;
<    scaledw = (int) ceil(kx*width_);
<    scaledh = (int) ceil(ky*height_);
< 
33a30
> 
53c50
< 	std::cerr << " PNGwriter::scale_kxky - ERROR **:  Not able to allocate memory for image." << std::endl;
---
> 	std::cerr << " PNGwriter::scale_wh - ERROR **:  Not able to allocate memory for image." << std::endl;
61c58
< 	     std::cerr << " PNGwriter::scale_kxky - ERROR **:  Not able to allocate memory for image." << std::endl;
---
> 	     std::cerr << " PNGwriter::scale_wh - ERROR **:  Not able to allocate memory for image." << std::endl;
67c64
< 	std::cerr << " PNGwriter::scale_kxky - ERROR **:  Not able to allocate memory for image." << std::endl;
---
> 	std::cerr << " PNGwriter::scale_wh - ERROR **:  Not able to allocate memory for image." << std::endl;
```
Note: So after a simple search and replace for `finale` -> `scaled`  we see that both functions do exactly the same with the difference of two or three lines precalculation.

**Comment by ax3l** [comment]
Thanks! :sparkles: Can you please summarize your changes in API in the PR description on top? The diff details below can be used for more detailed questions.

Please be aware we need  the `memset` sections since for black backgrounds (0 check) this improves the speed of the constructor tremendously. Please see #46 for comparison and check measure the change in init time.

**Comment by mxmlnkn** [comment]
The memset code wasn't deleted, see inside [createNewPngwriter](https://github.com/mxmlnkn/pngwriter/blob/60b0ed448f17bf43b7159f0f78f3de64257dbc3c/src/pngwriter.cc#L191-L196)

**Comment by ax3l** [comment]
Ok sounds good.

You got me wrong: it is not the `if` that slows things down. Please read #46 as linked.

**Comment by mxmlnkn** [comment]
My point is: I didn't change anything. The memset is still executed if the background color is black.

**Comment by ax3l** [comment]
Yep saw that, great work! :sparkles:

File: src/pngwriter.cc [hunk 1]
```diff
@@ -45,320 +45,32 @@
...
-	     graph_[vhhh][tempindex+2] = rhs.graph_[vhhh][tempindex+2];
-	     graph_[vhhh][tempindex+3] = rhs.graph_[vhhh][tempindex+3];
-	     graph_[vhhh][tempindex+4] = rhs.graph_[vhhh][tempindex+4];
-	     graph_[vhhh][tempindex+5] = rhs.graph_[vhhh][tempindex+5];
-	  }
-     }
-
+   /* the call to deleteMembers in the copy assignment method is no problem,
+    * because of the initializer list */
+   (*this) = rhs;
```
Comment (ax3l) [inline src/pngwriter.cc]: I am not aware that assigning directly to `*this` is allowed / a good idea in C++.

**Comment by ax3l** [comment]
I think the changes are so far ok.

Please update the following:
- do **not** change the constructors for now, please leave them as they are for now (also collides with #99)
- rebase against `dev`
- please check the functionality you changed (independent PR, maybe a **short** new test file)

**Pull request closed without merging**